File: apps/zitadel/zitadel-passkey-link.ts
#!/usr/bin/env bun
// Получить одноразовую ссылку для регистрации passkey (WebAuthn) у пользователя
// ZITADEL. Портирование zitadel/zitadel-passkey-link.sh на TypeScript.
//
// Зависимости — только Bun (нативные Bun.file и prompt):
//   bun zitadel/zitadel-passkey-link.ts
//
// Требует PAT администратора: переменная окружения ZITADEL_PAT или интерактивный ввод.

const scriptDir = import.meta.dir;
const repoRoot = `${scriptDir}/..`;

// Bun.color отдаёт пустую строку вне TTY — цвета сами отключаются при пайпе.
const paint = (color: string, msg: string) => console.error(`\x1b[1m${Bun.color(color, "ansi-16")}${msg}\x1b[0m`);
function info(msg: string) {
  paint("cyan", msg);
}
function ok(msg: string) {
  paint("green", msg);
}
function warn(msg: string) {
  paint("yellow", msg);
}
function die(msg: string): never {
  paint("red", `Ошибка: ${msg}`);
  process.exit(1);
}

const readEnvFile = async (path: string): Promise<Record<string, string>> => {
  const file = Bun.file(path);
  if (!(await file.exists())) return {};
  const entries: Record<string, string> = {};
  for (const line of (await file.text()).split("\n")) {
    const [, key, value] = line.match(/^([A-Za-z_][A-Za-z0-9_]*)=(.*)$/) ?? [];
    if (key && value) entries[key] = value.trim();
  }
  return entries;
};

// grpc-gateway отдаёт ошибку и с кодом 200 — общий фильтр по .message.
async function api(method: string, path: string, pat: string, apiBase: string, body?: unknown): Promise<any> {
  const resp = await fetch(`${apiBase}${path}`, {
    method,
    headers: {
      Authorization: `Bearer ${pat}`,
      ...(body !== undefined && { "Content-Type": "application/json" }),
    },
    body: body !== undefined ? JSON.stringify(body) : undefined,
  });
  const data: any = await resp.json();
  if (typeof data?.message === "string" && data.message !== "") die(data.message);
  return data;
}

function ask(question: string): string {
  const answer = prompt(question);
  return (answer ?? "").trim();
}

async function main() {
  const rootEnv = await readEnvFile(`${repoRoot}/.env`);
  const serviceEnv = await readEnvFile(`${scriptDir}/.env`);
  const domain = process.env.DOMAIN ?? rootEnv.DOMAIN ?? "example.com";
  const host = process.env.ZITADEL_HOST ?? serviceEnv.ZITADEL_HOST ?? `id.${domain}`;

  const apiBase = `https://${host}/v2`;
  const loginBase = `https://${host}/ui/v2/login`;

  let pat = process.env.ZITADEL_PAT ?? "";
  if (!pat) pat = ask("PAT (Console → Users → <admin> → Personal Access Tokens): ");
  if (!pat) die("PAT не задан (переменная ZITADEL_PAT или интерактивный ввод).");

  info(`ZITADEL: https://${host}`);

  const target = ask("Логин или user ID (например user или 386564404046479363): ");
  if (!target) die("пустой ввод.");

  // Один вызов ListUsers на оба случая: число — точный поиск по ID,
  // иначе — подстрока логина без учёта регистра.
  const query = /^\d{5,}$/.test(target)
    ? { inUserIdsQuery: { userIds: [target] } }
    : { loginNameQuery: { loginName: target, method: "TEXT_QUERY_METHOD_CONTAINS_IGNORE_CASE" } };
  const users: Array<[string, string, string, string]> = (await api("POST", "/users", pat, apiBase, { queries: [query] }))
    .result?.map((r: any) => [r.userId, r.username ?? "", r.preferredLoginName ?? "", r.details?.resourceOwner ?? ""])
    ?? [];

  if (users.length === 0) die(`пользователь по "${target}" не найден.`);

  let user: [string, string, string, string];
  if (users.length === 1) {
    const u = users[0];
    if (!u) die("не удалось прочитать результат поиска.");
    user = u;
  } else {
    info("Найдено несколько пользователей:");
    users.forEach(([uid, uname, ulogin], i) => console.error(`  ${i + 1}) ${uname} (${ulogin})  id=${uid}`));
    const idx = Number(ask("Номер: "));
    if (!Number.isInteger(idx) || idx < 1 || idx > users.length) die("неверный номер.");
    const u = users[idx - 1];
    if (!u) die("неверный номер.");
    user = u;
  }
  const [userId, username, loginName, orgId] = user;

  if (!userId || !orgId) die("не удалось определить user ID / organization ID.");

  info(`Пользователь: ${username || "<без username>"} (${loginName || "<без логина>"})`);
  info(`user_id=${userId}  organization=${orgId}`);

  const code = (await api("POST", `/users/${userId}/passkeys/registration_link`, pat, apiBase, { returnCode: {} })).code ?? {};
  if (!code.id || !code.code) die("в ответе API нет кода.");

  const url = `${loginBase}/passkey/set?` + new URLSearchParams({ codeId: code.id, code: code.code, userId, organization: orgId });

  ok("Ссылка для регистрации passkey:");
  console.log(`\x1b[1m${url}\x1b[0m`);
  console.log();
  warn(
    "Код одноразовый: ссылку надо открыть на том устройстве, где будет храниться passkey, " +
      "и пройти регистрацию до конца с первого раза (ZITADEL #12499).",
  );
}

await main();

File: apps/zitadel/zitadel-passkey-link.py
#!/usr/bin/env python3
#
# Получить одноразовую ссылку для регистрации passkey (WebAuthn) у пользователя
# ZITADEL. Портирование zitadel/zitadel-passkey-link.sh на Python (только
# стандартная библиотека, без зависимостей).
#
# Требует PAT администратора: переменная окружения ZITADEL_PAT или интерактивный ввод.
#
# Пример:
#   ZITADEL_PAT=... ./zitadel/zitadel-passkey-link.py

from __future__ import annotations

import json
import os
import re
import sys
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path

SCRIPT_DIR = Path(__file__).resolve().parent
REPO_ROOT = SCRIPT_DIR.parent


def info(msg: str) -> None:
    print(f"\033[1;36m{msg}\033[0m", file=sys.stderr)


def ok(msg: str) -> None:
    print(f"\033[1;32m{msg}\033[0m", file=sys.stderr)


def warn(msg: str) -> None:
    print(f"\033[1;33m{msg}\033[0m", file=sys.stderr)


def die(msg: str) -> "NoReturn":  # type: ignore[name-defined]
    print(f"\033[1;31mОшибка: {msg}\033[0m", file=sys.stderr)
    sys.exit(1)


def read_env_key(env_file: Path, key: str) -> str:
    if not env_file.is_file():
        return ""
    m = re.search(rf"^{key}=(.*)$", env_file.read_text(encoding="utf-8"), re.MULTILINE)
    return m.group(1).strip() if m else ""


def api(method: str, path: str, pat: str, api_base: str, body: dict | None = None) -> dict:
    req = urllib.request.Request(
        f"{api_base}{path}",
        method=method,
        headers={"Authorization": f"Bearer {pat}"},
        data=json.dumps(body).encode() if body is not None else None,
    )
    if body is not None:
        req.add_header("Content-Type", "application/json")
    try:
        with urllib.request.urlopen(req) as resp:
            data = json.load(resp)
    except urllib.error.HTTPError as e:
        die(str(json.load(e).get("message") or f"HTTP {e.code}"))
    # grpc-gateway отдаёт ошибку и с кодом 200 — общий фильтр по .message.
    if message := data.get("message"):
        die(message)
    return data


def main() -> None:
    domain = os.environ.get("DOMAIN") or "example.com"
    host = os.environ.get("ZITADEL_HOST") or f"id.{domain}"

    api_base = f"https://{host}/v2"
    login_base = f"https://{host}/ui/v2/login"

    pat = os.environ.get("ZITADEL_PAT") or ""
    if not pat:
        import getpass

        try:
            pat = getpass.getpass("PAT (Console → Users → <admin> → Personal Access Tokens): ")
        except EOFError:
            pass
    if not pat:
        die("PAT не задан (переменная ZITADEL_PAT или интерактивный ввод).")

    info(f"ZITADEL: https://{host}")

    target = input("Логин или user ID (например user или 386564404046479363): ").strip()
    if not target:
        die("пустой ввод.")

    # Один вызов ListUsers на оба случая: число — точный поиск по ID,
    # иначе — подстрока логина без учёта регистра.
    if re.fullmatch(r"\d{5,}", target):
        query: dict = {"inUserIdsQuery": {"userIds": [target]}}
    else:
        query = {
            "loginNameQuery": {
                "loginName": target,
                "method": "TEXT_QUERY_METHOD_CONTAINS_IGNORE_CASE",
            }
        }
    users = [
        (
            r["userId"],
            r.get("username", ""),
            r.get("preferredLoginName", ""),
            r.get("details", {}).get("resourceOwner", ""),
        )
        for r in api("POST", "/users", pat, api_base, {"queries": [query]}).get("result", [])
    ]

    if not users:
        die(f'пользователь по "{target}" не найден.')

    if len(users) == 1:
        user_id, username, login, org_id = users[0]
    else:
        info("Найдено несколько пользователей:")
        for i, (uid, uname, ulogin, _) in enumerate(users, start=1):
            print(f"  {i}) {uname} ({ulogin})  id={uid}", file=sys.stderr)
        choice = input("Номер: ").strip()
        if not choice.isdigit() or not 1 <= int(choice) <= len(users):
            die("неверный номер.")
        user_id, username, login, org_id = users[int(choice) - 1]

    if not user_id or not org_id:
        die("не удалось определить user ID / organization ID.")

    info(f"Пользователь: {username or '<без username>'} ({login or '<без логина>'})")
    info(f"user_id={user_id}  organization={org_id}")

    code = api(
        "POST",
        f"/users/{user_id}/passkeys/registration_link",
        pat,
        api_base,
        {"returnCode": {}},
    ).get("code", {})
    code_id, code_value = code.get("id", ""), code.get("code", "")
    if not code_id or not code_value:
        die("в ответе API нет кода.")

    url = (
        f"{login_base}/passkey/set?"
        + urllib.parse.urlencode(
            {"codeId": code_id, "code": code_value, "userId": user_id, "organization": org_id}
        )
    )

    ok("Ссылка для регистрации passkey:")
    print(f"\033[1m{url}\033[0m")
    print()
    warn(
        "Код одноразовый: ссылку надо открыть на том устройстве, где будет храниться "
        "passkey, и пройти регистрацию до конца с первого раза (ZITADEL #12499)."
    )


if __name__ == "__main__":
    main()
